src/Provider/DefaultSmsProvider.php

```diff
@@ -110,23 +114,24 @@ public function send(SmsMessageInterface $sms) {
    * {@inheritdoc}
    */
   public function incoming(SmsMessageInterface $sms_message) {
+    $sms_message->setDirection(Direction::INCOMING);
+
+    // Do not iterate over messages individually like outgoing, changing gateway
+    // in pre-process events do not apply to incoming.
+    $plugin = $sms_message->getGateway()->getPlugin();
+
     $dispatch = !$sms_message->getOption('_skip_preprocess_event');
     $sms_messages = $dispatch ? $this->dispatchEvent(SmsEvents::MESSAGE_PRE_PROCESS, [$sms_message])->getMessages() : [$sms_message];
     $sms_messages = $this->dispatchEvent(SmsEvents::MESSAGE_INCOMING_PRE_PROCESS, $sms_messages)->getMessages();
 
-    foreach ($sms_messages as &$sms_message) {
-      $plugin = $sms_message->getGateway()->getPlugin();
-      if (!$plugin instanceof SmsGatewayPluginIncomingInterface) {
-        throw new SmsException(sprintf('Gateway plugin `%s` does not support incoming messages', $plugin->getPluginId()));
-      }
-
-      $result = $plugin->incoming($sms_message);
-      $sms_message->setResult($result);
-
-      $this->dispatchEvent(SmsEvents::MESSAGE_INCOMING_POST_PROCESS, [$sms_message]);
-      $this->dispatchEvent(SmsEvents::MESSAGE_POST_PROCESS, [$sms_message]);
+    if ($plugin instanceof SmsGatewayPluginIncomingEventInterface) {
```

Comment:
This interface is designed specifically for gateway plugins, do we not want to prefix these interfaces with `SmsGateway` or `SmsGatewayPlugin`?

- `SmsGatewayPluginIncomingEventInterface` (current, prefixed)
- `SmsGatewayPluginIncomingEventSubscriberInterface`
- `SmsGatewayPluginIncomingEventProcessorInterface` I'd probably go with this one.

Bit of a mouthful 😒